File: oiak_project/project/algorithms/SystemConversion.cpp

#include "SystemConversion.h"

SystemConversion::SystemConversion() {}
SystemConversion::~SystemConversion() {}
using namespace std;
string SystemConversion::binToHex(int &number) {
    int hexValue = 0, decValue = 0;
    decValue = binToDec(number);
    string myHex, hexResult;
    while (decValue != 0)
    {
        hexValue = decValue % 16;
        if(hexValue > 9)
        {
            myHex.push_back((char)(hexValue + 55));
        }
        else
        {
            myHex.push_back((char)(hexValue + 48));
        }
        decValue = decValue / 16;
    }

    for(int j = myHex.size() - 1; j >= 0; j--)
    {
        hexResult.push_back(myHex[j]);
    }
    return hexResult;
}

void SystemConversion::runBinToHex() {
    cout << "Welcome to the binary to hexadecimal converter\n";
    string binaryNum;
    binaryNum = "2";
    while(!isBinary(binaryNum))
    {
        cout << "Enter a binary number\n";
        cin >> binaryNum;
        if(!isBinary(binaryNum))
        {
            cout << "Invalid value\n";
        }
    }
    int binNum;
    binNum = stringToInt(binaryNum);
    cout << "Result = " << binToHex(binNum) << endl;

}
void SystemConversion::runBinToDec() {
    cout << "Welcome to the binary to decimal converter\n";
    string binaryNum;
    binaryNum = "2";
    while(!isBinary(binaryNum))
    {
        cout << "Enter a binary number\n";
        cin >> binaryNum;
        if(!isBinary(binaryNum))
        {
            cout << "Invalid value\n";
        }
    }
    int binNum;
    binNum = stringToInt(binaryNum);
    cout << "Result = " << binToDec(binNum) << endl;

}
void SystemConversion::runHexToBin() {
    cout << "Welcome to the hexadecimal to binary converter\n";
    string hexNum;
    hexNum = "t";
    while(!isHex(hexNum))
    {
        cout << "Enter a hexadecimal number\n";
        cin >> hexNum;
        if(!isHex(hexNum))
        {
            cout << "Invalid value\n";
        }
    }
    cout << "Result = " << hexToBin(hexNum) << endl;

}

int SystemConversion::binToDec(int &number) {
    int decValue = 0, i = 1, r;
    while (number > 0)
    {
        r = number % 2;
        decValue = decValue + r * i;
        i = i * 2;
        number = number / 10;
    }
    return decValue;

}
string SystemConversion::hexToBin(string &number) {
    int i = 0;
    string binaryNum;
    while (number[i]) {

        switch (number[i]) {
            case '0':
                binaryNum = binaryNum + "0000";
                break;
            case '1':
                binaryNum = binaryNum + "0001";
                break;
            case '2':
                binaryNum = binaryNum + "0010";
                break;
            case '3':
                binaryNum = binaryNum + "0011";
                break;
            case '4':
                binaryNum = binaryNum + "0100";
                break;
            case '5':
                binaryNum = binaryNum + "0101";
                break;
            case '6':
                binaryNum = binaryNum + "0110";
                break;
            case '7':
                binaryNum = binaryNum + "0111";
                break;
            case '8':
                binaryNum = binaryNum + "1000";
                break;
            case '9':
                binaryNum = binaryNum + "1001";
                break;
            case 'A':
            case 'a':
                binaryNum = binaryNum + "1010";
                break;
            case 'B':
            case 'b':
                binaryNum = binaryNum + "1011";
                break;
            case 'C':
            case 'c':
                binaryNum = binaryNum + "1100";
                break;
            case 'D':
            case 'd':
                binaryNum = binaryNum + "1101";
                break;
            case 'E':
            case 'e':
                binaryNum = binaryNum + "1110";
                break;
            case 'F':
            case 'f':
                binaryNum = binaryNum + "1111";
                break;
            default:
                cout << "Invalid value\n";
        }
        i++;
    }
    return binaryNum;

}

int SystemConversion::hexToDec(string &number) {
    int len = number.size();
    int base = 1;
    int decValue = 0;
    for (int i = len - 1; i >= 0; i--) {
        if (number[i] >= '0' && number[i] <= '9') {
            decValue += (int(number[i]) - 48) * base;
            base = base * 16;
        }
        else if (number[i] >= 'A' && number[i] <= 'F') {
            decValue += (int(number[i]) - 55) * base;
            base = base * 16;
        }
    }
    return decValue;
}

void SystemConversion::runHexToDec() {
    cout << "Welcome to the hexadecimal to decimal converter\n";
    string hexNum;
    hexNum = "t";
    while(!isHex(hexNum))
    {
        cout << "Enter a hexadecimal number\n";
        cin >> hexNum;
        if(!isHex(hexNum))
        {
            cout << "Invalid value\n";
        }
    }
    cout << "Result = " << hexToDec(hexNum) << endl;

}

int SystemConversion::decToBin(int &number) {
    int binary = 0, r, tmp = 1;
    while (number != 0) {
        r = number % 2;
        binary = binary + (r * tmp);
        number = number / 2;
        tmp *= 10;
    }
    return binary;
}
string SystemConversion::decToHex(int &number) {
    int r;
    string result;
    while (number > 0) {
        r = number % 16;
        if (r > 9) {
            switch (r) {
                case 10:
                    result = result + "A";
                    break;
                case 11:
                    result = result + "B";
                    break;
                case 12:
                    result = result + "C";
                    break;
                case 13:
                    result = result + "D";
                    break;
                case 14:
                    result = result + "E";
                    break;
                case 15:
                    result = result + "F";
                    break;
            }
        }
        else {
            result.push_back(char(r + 48));
        }
        number = number / 16;
    }
    reverse(result.begin(), result.end());
    return result;
}

void SystemConversion::runDecToHex() {
    std::cout << "Welcome to the decimal to binary converter\n";
    inputNumber = -1;
    while(inputNumber < 0)
    {
        std::cout << "Enter positive int number\n";
        inputNumber = getPositiveNumber();
        if(inputNumber < 0)
        {
            std::cout << "Invalid value\n";
        }
    }
    cout << "Result = " << decToHex(inputNumber) << endl;
}
void SystemConversion::runDecToBin() {
    std::cout << "Welcome to the decimal to binary converter\n";
    inputNumber = -1;
    while(inputNumber < 0)
    {
        std::cout << "Enter positive int number\n";
        inputNumber = getPositiveNumber();
        if(inputNumber < 0)
        {
            std::cout << "Invalid value\n";
        }
    }
    cout << "Result = " << decToBin(inputNumber) << endl;
}
bool SystemConversion::isBinary(string &number) {
    for(int i = 0 ; i < number.size() ; i++)
        if(number[i] > '1') {
            return false;
        }
    return true;
}
bool SystemConversion::isHex(string &number) {
    for(int i = 0; i < number.size(); i++)
    {
        if(!isxdigit(number[i]))
        {
            return false;
        }
    }
    return true;
}
int SystemConversion::getPositiveNumber() {
    int option = -1;
    cin >> option;
    if(cin.fail()){
        cin.clear();
        cin.ignore();
        return -1;
    }
        if(option < 0)
        {
            return -1;
        }
        return option;
}

int SystemConversion::stringToInt(string &number) {
    return stoi(number);
}

File: python/system_conversion.py
import random
import string

class SystemConversion:
    def __init__(self, number):
        self.number = number

    def dec_to_oct(dec_num):
        result =[]
        r=0
        number=dec_num
        while(number!=0):
            r=number%8
            number=number//8
            _r=str(r)
            result.append(_r)
        result.reverse()
        return ''.join(result)

    def dec_to_bin(dec_num):
        result = []
        r=0
        number=dec_num

        while(number!=0):
            r=number%2
            number=number//2
            _r=str(r)
            result.append(_r)
        
        result.reverse()
        return ''.join(result)

    def dec_to_hex(dec_num)->string:     
        result = []
        r=0
        number = dec_num
        while(number!=0):
            r=number % 16
            number = number//16
            # jesli reszta z dzielenia wieksza niz 9, dodaj litere to wyniku
            if(r>9):
                match r:
                    case 10:
                        result.append('A')
                    case 11:
                        result.append('B')
                    case 12:
                        result.append('C')
                    case 13:
                        result.append('D')
                    case 14:
                        result.append('E')
                    case 15:
                        result.append('F')
            else:
                _r=str(r)
                result.append(_r)
        result.reverse()
        result2 = ''.join(result)
        return result2

    def bin_to_dec(bin_num)->int:
        r=0
        result=0
        weight=1
        while(bin_num!=0):
            r=bin_num%2
            result =result + r*weight
            weight=2*weight
            bin_num=bin_num//10
        return result

    def bin_to_hex(bin_num)->string:
        number=SystemConversion.bin_to_dec(bin_num)
        result=SystemConversion.dec_to_hex(number)
        return result

    def hex_to_bin(hex_num)->string:
        result = []
        for i in range(len(hex_num)):
            match hex_num[i]:
                case '0':
                    result.append('0000')
                case '1':
                    result.append('0001')
                case '2':
                    result.append('0010')
                case '3':
                    result.append('0011')
                case '4':
                    result.append('0100')
                case '5':
                    result.append('0101')
                case '6':
                    result.append('0110')
                case '7':
                    result.append('0111')
                case '8':
                    result.append('1000')
                case '9':
                    result.append('1001')
                case 'A':
                    result.append('1010')
                case 'B':
                    result.append('1011')
                case 'C':
                    result.append('1100')
                case 'D':
                    result.append('1101')
                case 'E':
                    result.append('1110')
                case 'F':
                    result.append('1111')
        return ''.join(result)

    def hex_to_dec(hex_num):
        _number=SystemConversion.hex_to_bin(hex_num)
        number=int(_number)
        result=SystemConversion.bin_to_dec(number)
        return result